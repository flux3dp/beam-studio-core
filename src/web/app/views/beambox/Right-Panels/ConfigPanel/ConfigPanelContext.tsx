import { createContext, Dispatch } from 'react';

import { DataType, dataKey, defaultConfig } from 'helpers/layer/layer-config-helper';
import { ILayerConfig } from 'interfaces/ILayerConfig';

interface State extends ILayerConfig {
  selectedItem?: string;
}

export const getDefaultState = (): State => {
  const keys = Object.keys(defaultConfig);
  const initState = {} as State;
  keys.forEach((key) => {
    // Handle DataType and state type mismatch
    initState[dataKey[key]] = { value: defaultConfig[key as DataType] };
  });
  return initState;
};

export type Action = {
  type: 'update';
  payload: State;
} | {
  type: 'change';
  payload: {
    selectedItem?: string;
  } & {
    [key in keyof ILayerConfig]?: number | string;
  };
} | {
  type: 'rename';
  payload: string;
};

export const reducer = (state: State, action: Action): State => {
  if (action.type === 'update') return { ...state, ...action.payload };
  if (action.type === 'change') {
    const { payload } = action;
    const newState = { ...state };
    Object.keys(payload).forEach((key) => {
      if (key !== 'selectedItem') newState[key] = { value: payload[key] };
      else newState[key] = payload[key];
    });
    return newState;
  }
  if (action.type === 'rename') {
    const { payload } = action;
    const newState = { ...state };
    newState.configName = { value: payload };
    newState.selectedItem = payload;
    return newState;
  }
  return state;
};

interface Context {
  state: State;
  selectedLayers: string[];
  dispatch: Dispatch<Action>;
  simpleMode?: boolean;
  initState: (layers?: string[]) => void;
}

export default createContext<Context>({} as Context);
