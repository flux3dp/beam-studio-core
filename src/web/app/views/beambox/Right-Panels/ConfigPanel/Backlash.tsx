import React, { memo, useContext } from 'react';

import ConfigSlider from 'app/views/beambox/Right-Panels/ConfigPanel/ConfigSlider';
import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';
import { DataType, writeData } from 'helpers/layer/layer-config-helper';

import ConfigPanelContext from './ConfigPanelContext';
import styles from './Block.module.scss';

const Backlash = (): JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;

  const { selectedLayers, state, dispatch } = useContext(ConfigPanelContext);
  const { backlash } = state;
  const handleChange = (value: number) => {
    dispatch({
      type: 'change',
      payload: { backlash: value },
    });
    selectedLayers.forEach((layerName) => {
      writeData(layerName, DataType.backlash, value);
    });
  };

  return (
    <div className={styles.panel}>
      <span className={styles.title}>{t.backlash}</span>
      <UnitInput
        id="backlash"
        className={{ [styles.input]: true }}
        min={0}
        max={10}
        unit="mm"
        defaultValue={backlash.value}
        getValue={handleChange}
        decimal={2}
        displayMultiValue={backlash.hasMultiValue}
      />
      <ConfigSlider
        id="backlash"
        value={backlash.value}
        onChange={handleChange}
        min={0}
        max={10}
        step={0.1}
      />
    </div>
  );
};

export default memo(Backlash);
