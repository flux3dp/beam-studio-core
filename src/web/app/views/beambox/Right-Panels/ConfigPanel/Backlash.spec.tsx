import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import Backlash from './Backlash';
import ConfigPanelContext from './ConfigPanelContext';

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    right_panel: {
      laser_panel: {
        backlash: 'backlash',
      },
    },
  },
}));

jest.mock('app/widgets/Unit-Input-v2', () => (
  { id, min, max, unit, defaultValue, decimal, displayMultiValue }: any
) => (
  <div>
    MockUnitInput
    <p>id: {id}</p>
    <p>min: {min}</p>
    <p>max: {max}</p>
    <p>unit: {unit}</p>
    <p>defaultValue: {defaultValue}</p>
    <p>decimal: {decimal}</p>
    <p>displayMultiValue: {displayMultiValue}</p>
  </div>
));

const mockWriteData = jest.fn();
jest.mock('helpers/layer/layer-config-helper', () => ({
  DataType: {
    backlash: 'backlash',
  },
  writeData: (...args) => mockWriteData(...args),
}));

const mockSelectedLayers = ['layer1', 'layer2'];
const mockContextState = {
  backlash: { value: 8.7, hasMultiValue: false },
};
const mockDispatch = jest.fn();

describe('test Backlash', () => {
  it('should render correctly', () => {
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{ state: mockContextState as any, dispatch: mockDispatch, selectedLayers: mockSelectedLayers }}
      >
        <Backlash />
      </ConfigPanelContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('onChange should work', () => {
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{ state: mockContextState as any, dispatch: mockDispatch, selectedLayers: mockSelectedLayers }}
      >
        <Backlash />
      </ConfigPanelContext.Provider>
    );
    expect(mockDispatch).not.toBeCalled();
    expect(mockWriteData).not.toBeCalled();
    const input = container.querySelector('input');
    fireEvent.change(input, { target: { value: '8.8' } });
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({
      type: 'change',
      payload: { backlash: 8.8 },
    });
    expect(mockWriteData).toBeCalledTimes(2);
    expect(mockWriteData).toHaveBeenNthCalledWith(1, 'layer1', 'backlash', 8.8);
    expect(mockWriteData).toHaveBeenNthCalledWith(2, 'layer2', 'backlash', 8.8);
  });
});
