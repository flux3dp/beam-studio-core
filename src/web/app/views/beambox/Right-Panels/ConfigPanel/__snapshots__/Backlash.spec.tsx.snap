// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Backlash should render correctly 1`] = `
<div>
  <div
    class="panel"
  >
    <span
      class="title"
    >
      backlash
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        backlash
      </p>
      <p>
        min: 
        0
      </p>
      <p>
        max: 
        10
      </p>
      <p>
        unit: 
        mm
      </p>
      <p>
        defaultValue: 
        8.7
      </p>
      <p>
        decimal: 
        2
      </p>
      <p>
        displayMultiValue: 
      </p>
    </div>
    <input
      max="10"
      min="0"
      step="0.1"
      type="range"
      value="8.7"
    />
  </div>
</div>
`;
